src/Stratis.Bitcoin.Features.BlockStore/BlockStore.cs

```diff
@@ -0,0 +1,363 @@
+                    break;
+                }
+
+                resetBlock = await this.blockRepository.GetAsync(resetBlock.Header.HashPrevBlock).ConfigureAwait(false);
+                Guard.NotNull(resetBlock, nameof(resetBlock));
+                resetBlockHash = resetBlock.GetHash();
+            }
+
+            ChainedHeader newTip = this.chain.GetBlock(resetBlockHash);
+            await this.blockRepository.DeleteAsync(newTip.HashBlock, blockStoreResetList).ConfigureAwait(false);
+
+            this.SetStoreTip(newTip);
+
+            this.logger.LogWarning("Block store tip recovered to block '{0}'.", newTip);
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// Adds a block to the saving queue.
+        /// </summary>
+        /// <param name="blockPair">The block and its chained header pair to be added to pending storage.</param>
+        public void AddToPending(BlockPair blockPair)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(blockPair), blockPair.ChainedHeader);
+
+            this.blocksQueue.Enqueue(blockPair);
+
+            this.logger.LogTrace("(-)");
+        }
+        
+        /// <summary>
+        /// Dequeues the blocks continuously and saves them to the database when max batch size is reached or timer ran out.
+        /// </summary>
+        /// <remarks>Batch is always saved on shutdown.</remarks>
+        private async Task DequeueBlocksContinuouslyAsync()
+        {
+            this.logger.LogTrace("()");
+
+            var batch = new List<BlockPair>();
+
+            Task<BlockPair> dequeueTask = null;
+            Task timerTask = null;
+
+            while (!this.nodeLifetime.ApplicationStopping.IsCancellationRequested)
+            {
+                // Start new dequeue task if not started already.
+                dequeueTask = dequeueTask ?? this.blocksQueue.DequeueAsync();
+
+                // Wait for one of the tasks: dequeue or timer (if available) to finish.
+                Task task = (timerTask == null) ? dequeueTask : await Task.WhenAny(dequeueTask, timerTask).ConfigureAwait(false);
+
+                bool saveBatch = false;
+
+                try
+                {
+                    await task.ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    // Happens when node is shutting down or Dispose() is called.
+                    // We want to save whatever is in the batch before exiting the loop.
+                    saveBatch = true;
+
+                    this.logger.LogDebug("Node is shutting down. Save batch.");
+                }
+
+                // Save batch if timer ran out or we've dequeued a new block and reached the consensus tip 
+                // or the max batch size is reached or the node is shutting down.  
+                if (dequeueTask.Status == TaskStatus.RanToCompletion)
```

Comment:
this feels worthy of its own method perhaps - quite a few things being checked and set

Reply:
This one mostly uses lots of private members so it doesn't really make sense to move it to a separated method- just adds more complexity for the reader